### PR TITLE
Add script for detecting bad characters.

### PR DESCRIPTION
**Pull request by s-kawamura-w664**
This is script that find control-code(0x00-0x1f) in text files.
I could find them in several files. Please look at PR #26965, #27173, #27171.

### CODE REVIEW

**Review by sftim — COMMENTED** [review]
Nice idea

We could run this as a CI test, in fact.

**Comment by s-ito-ts** [comment]
Could you add how to use this script to the README?
https://github.com/kubernetes/website/blob/master/scripts/README.md

**Comment by oomichi** [comment]
> We could run this as a CI test, in fact.

@sftim That is a nice idea.
Do you know how to enable some script on CI test of this website repo?
On kubernetes/kubernetes repo, hack/verify-***.sh files are distinguished as CI testing scripts and these scripts are automatically operated. But I could not find such way on this repo.

/cc @oomichi

**Comment by s-kawamura-w664** [comment]
@s-ito-ts 
I updated README.md.
Could you please review?

> Could you add how to use this script to the README?
> https://github.com/kubernetes/website/blob/master/scripts/README.md

File: scripts/check-ctrlcode.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    fullpath = os.path.abspath(dirpath)
+    if (os.path.isdir(fullpath) is not True):
+        print("Directory not found.")
+        sys.exit(1)
+
+    check_dir(fullpath, ext)
+
+def check_dir(path, ext):
+    for f in os.listdir(path):
+        if(f[0] != "."):
```
Comment (oomichi) [inline scripts/check-ctrlcode.py:25]: To reduce nests, here can be
```
if(f[0] == "."):
    continue
fullpath = os.path.join(path, f)
...
```

File: scripts/check-ctrlcode.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+            fullpath = os.path.join(path, f)
+            if(os.path.isfile(fullpath)):
+                exts = os.path.splitext(f)
+                if(exts[1] == ext):
```
Comment (oomichi) [inline scripts/check-ctrlcode.py:29]: ditto:
```
if(exts[1] != ext):
    continue
check_ctrlcode(fullpath)
```

File: scripts/check-ctrlcode.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                    check_ctrlcode(fullpath)
+            else:
+                check_dir(fullpath, ext)
+
+def check_ctrlcode(filepath):
+    line = 0
+    with open(filepath, encoding='utf-8') as f:
+        while True:
+            str = f.readline()
+            if str:
```
Comment (oomichi) [inline scripts/check-ctrlcode.py:39]: ditto:
```
if str == "":
    break
line = line + 1
# check 0x00-0x1f except 0x09(HT), 0x0a(LF), 0x0d(CR)
...
```

File: scripts/check-ctrlcode.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                line = line + 1
+                # check 0x00-0x1f except 0x09(HT), 0x0a(LF), 0x0d(CR)
+                pattern = re.compile('[\u0000-\u0008\u000b\u000c\u000e-\u001f]')
+                m = pattern.search(str)
+                if(m != None):
```
Comment (oomichi) [inline scripts/check-ctrlcode.py:44]: ditto:
```
if(m == None):
    continue
pos = m.end()
ctrl = m.group().encode("utf-8")
...
```

**Review by oomichi — COMMENTED** [review]
some comments to make the code more readable.

**Comment by s-kawamura-w664** [comment]
> some comments to make the code more readable.

@oomichi 
I fixed them according to your review comment.
Could you please review my fixes?

File: scripts/check-ctrlcode.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        if(f[0] == "."):
+            continue
+        fullpath = os.path.join(path, f)
+        if(os.path.isfile(fullpath)):
+            exts = os.path.splitext(f)
+            if(exts[1] != ext):
+                continue
+            check_ctrlcode(fullpath)
+        else:
+            check_dir(fullpath, ext)
```
Comment (oomichi) [inline scripts/check-ctrlcode.py:34]: similar point. How about changing line 30-34 to
```
if(os.path.isdir(fullpath)):
    check_dir(fullpath, ext)
    continue
exts = os.path.splitext(f)
if(exts[1] != ext):
    continue
check_ctrlcode(fullpath)
```
?

**Review by oomichi — COMMENTED** [review]
Thanks for updating.
Most part seems good for me, just one comment.

File: scripts/check-ctrlcode.py (hunk 5 above)
Comment (s-kawamura-w664) [reply scripts/check-ctrlcode.py:34]: I don't think nest is a problem.
I thought about the ease of understanding the conditions.
My code feels easy to understand the if-else condition.
Your code feels easy to understand as a loop condition.
I would like to accept your proposal and fix it.
After fixed, I'll request you to review.

Comment (oomichi) [reply scripts/check-ctrlcode.py:34]: This point is not only for nest.
This point makes condition same as the next called function.
In my suggestion, `isdir()` condition calls `check_dir()`.
I think that is easy to understand what is the purpose here.

**Comment by s-kawamura-w664** [comment]
> Thanks for updating.
> Most part seems good for me, just one comment.

@oomichi 
Thanks for your comment.
I fixed it.
Could you please review my fixes?

**Review by oomichi — COMMENTED** [review]
Thanks for updating.

/lgtm

**Comment by s-kawamura-w664** [comment]
/assign @steveperry-53

Could you please review?
If this is OK, please approve.

**Review by savitharaghunathan — COMMENTED** [review]
/approve

**Pull request merged**